src/main/java/net/neoforged/neoforge/client/util/TooltipUtil.java

```diff
@@ -0,0 +1,171 @@
+
+            if (modifierMap.isEmpty()) return;
+
+            Map<Holder<Attribute>, BaseModifier> baseModifs = new IdentityHashMap<>();
+
+            modifierMap.forEach((attr, modif) -> {
+                if (modif.id().equals(attr.value().getBaseID())) {
+                    baseModifs.put(attr, new BaseModifier(modif, new ArrayList<>()));
+                }
+            });
+
+            modifierMap.forEach((attr, modif) -> {
+                BaseModifier base = baseModifs.get(attr);
+                if (base != null && base.base != modif) {
+                    base.children.add(modif);
+                }
+            });
+
+            for (Map.Entry<Holder<Attribute>, BaseModifier> entry : baseModifs.entrySet()) {
+                Holder<Attribute> attr = entry.getKey();
+                BaseModifier baseModif = entry.getValue();
+                double entityBase = player == null ? 0 : player.getAttributeBaseValue(attr);
+                double base = baseModif.base.amount() + entityBase;
+                final double rawBase = base;
+                double amt = base;
+                double baseBonus = attr.value().getBonusBaseValue(stack);
+                for (AttributeModifier modif : baseModif.children) {
+                    if (modif.operation() == Operation.ADD_VALUE) base = amt = amt + modif.amount();
+                    else if (modif.operation() == Operation.ADD_MULTIPLIED_BASE) amt += modif.amount() * base;
+                    else amt *= 1 + modif.amount();
+                }
+                amt += baseBonus;
+                boolean isMerged = !baseModif.children.isEmpty() || baseBonus != 0;
+                MutableComponent text = attr.value().toBaseComponent(amt, entityBase, isMerged, flag);
+                tooltip.accept(padded(" ", text).withStyle(isMerged ? ChatFormatting.GOLD : ChatFormatting.DARK_GREEN));
+                if (Screen.hasShiftDown() && isMerged) {
+                    // Display the raw base value, and then all children modifiers.
+                    text = attr.value().toBaseComponent(rawBase, entityBase, false, flag);
+                    tooltip.accept(list().append(text.withStyle(ChatFormatting.DARK_GREEN)));
+                    for (AttributeModifier modifier : baseModif.children) {
+                        tooltip.accept(list().append(attr.value().toComponent(modifier, flag)));
+                    }
+                    if (baseBonus > 0) {
+                        attr.value().addBonusTooltips(stack, tooltip, flag);
+                    }
+                }
+            }
+
+            for (Holder<Attribute> attr : modifierMap.keySet()) {
+                if (baseModifs.containsKey(attr)) continue;
+                Collection<AttributeModifier> modifs = modifierMap.get(attr);
+                // Initiate merged-tooltip logic if we have more than one modifier for a given attribute.
```

Comment:
Some discussion might be necessary to determine if NeoForge should merge same-operation-same-attribute modifiers, only, or merge modifiers of the same attribute across different operations.

Reply:
Merge logic has been locked behind the mod-enableable flag `NeoForgeMod.enableMergedAttributeTooltips()`